### ISSUE TITLE
No conforming way to opt out of layout table heurisitics

### ISSUE THREAD

**Issue by dd8** [w3c/html-aam#293]
![](https://pubchem.ncbi.nlm.nih.gov/periodic-table/png/Periodic_Table_of_Elements_w_Chemical_Group_Block_PubChem.png)
The HTML Standard encourages user agents to implement layout table detection heuristics, but only makes broad suggestions on what they should be. 

> This specification does not define a precise heuristic, but the following are suggested as possible indicators

https://html.spec.whatwg.org/multipage/tables.html#the-table-element

The spec has encouraged layout table detection heuristics for a long time - for at least 10 years:
https://www.w3.org/TR/2010/WD-html5-20100304/tabular-data.html#the-table-element

User agents *have* implemented heuristics but these are mostly not interoperable. There's a simplified summary of different heuristics used here:
https://www.powermapper.com/blog/layout-tables-vs-data-tables/#heuristics-by-user-agent

The HTML Standard essentially suggests tables have 2 implicit roles (data tables are equivalent to role=table and layout tables are equivalent to role=none) but html-aam always maps to role=table.

Marking a table as layout with `role=presentation` is interoperable and is conforming
Marking a table as data with `role=table` is partially interoperable but is not conforming
BUT there's no conforming way to mark all possible tables containing data as data tables (because adding `th` headers is interoperable but isn't an option for some data tables, and is itself a heuristic)

Problems with current situation:

- VoiceOver's heuristic ignores `th` elements unless they're in first row or column
- JAWS layout table heuristic very different to other browsers and most of the suggestions in the HTML Standard https://doccenter.freedomscientific.com/doccenter/doccenter/rs11f929e9c511/2012-05-08_teachersandtrainers-l11/02_tablesandforms.htm
- Soduko grids have row and column relationships, but no headers
- The simplest form of the Periodic Table used in introductory chemistry has no headers:
![Periodic Table of the Elements - irregular table with rows and columns based on physical reality](https://user-images.githubusercontent.com/16352274/85876678-fd0eb680-b7cd-11ea-95c5-17693c9f8887.png)

Related: https://github.com/w3c/html-aam/issues/83

**Comment by JAWS-test** [comment]
Your suggestion is to allow role=table on the table element? I think this is good as an interim solution, but in the long run it would be better to abolish the absurd table heuristics completely

**Comment by dd8** [comment]
Yes, I think the options are:

1) make `role=table` on `table` elements conforming and specify that it overrides the layout table heuristic
2) keep the status quo and accept tables aren't always voiced as tables
3) agree and specify layout table heuristics and wait for all AT/browser vendors to implement them
4) remove the layout table heuristics from the HTML spec and wait for all AT/browser vendors remove them

Allowing `role=table` on the `table` element (option 1) seems the easiest and lowest risk option since it's fairly rare in the wild and few authors will be surprised when `role=table` is voiced as a table. It also requires changing fewer specs (probably just html-aam) but options 3 and 4 likely require changes to the HTML Standard as well.

One downside is consistency - ARIA roles matching implicit roles on other elements is non-conforming, but as discussed above tables have two implicit roles due to layout table heuristics.

Removing or changing the heuristics might be a hard sell for AT vendors. The JAWS heuristic must cause lots of problems, but changing it risks breaking backwards compatibility for JAWS users in hard to predict ways. But I'd agree removing them longer term is a good aim.

There was some discussion on `role=table` here:
https://codepen.io/stevef/post/short-note

**Comment by dd8** [comment]
Here's a good example of layout table heuristics causing problems on a table with `th` headers - the Periodic Table on Wikipedia:
https://en.wikipedia.org/wiki/Periodic_table#Overview

- VoiceOver with Safari 13.1.1 - not read as table and no table semantics exposed in macOS Accessibility Inspector
- Chrome 83 - exposed as table in Accessibility tab in devtools
- Firefox 72 - exposed as table in Accessibility tab in devtools
- IE 11 with NVDA 2019.2 - voiced as a table
- Firefox 72 with NVDA 2019.2 - voiced as a table
- IE 11 with JAWS 2019 - voiced as a table
- Firefox 72 with JAWS 2019 - voiced as a table

Likely cause of the problem for VoiceOver is the gap in the headers between columns 3 and 4 representing the lanthanides and actinides. These shouldn't have a column name  - see https://en.wikipedia.org/wiki/Periodic_table#Groups for an explanation.

**tl;dr** using `th` doesn't guarantee a table will be treated as a data table

**Comment by JAWS-test** [comment]
I agree that role=table on the table element is currently the best solution

**Comment by scottaohara** [comment]
that table isn't exposed as a table in safari because of the use of `display: block`.  you remove that, it's a table again.

**Comment by scottaohara** [comment]
steve referenced https://html.spec.whatwg.org/multipage/tables.html#the-table-element in the previous issue.

per my above comment, browsers can perform their own undocumented magic when using `display: something` on a `<table>` element.

**Comment by zcorpan** [comment]
> Yes, I think the options are:
> 
> 1. make `role=table` on `table` elements conforming and specify that it overrides the layout table heuristic
> 2. keep the status quo and accept tables aren't always voiced as tables
> 3. agree and specify layout table heuristics and wait for all AT/browser vendors to implement them
> 4. remove the layout table heuristics from the HTML spec and wait for all AT/browser vendors remove them

(2) is not great.

I think (4) is not great for users -- the likely effect would be they would be a lot of layout tables get exposed as data tables.

(3) seems like a good thing to me, getting the heuristic more interoperable and predictable means more consistent experience for users across browsers and ATs, and authors have higher confidence in how their markup will be presented in ATs.

It doesn't exclude doing (1). It also seems like a good thing to me. For elements that have heuristics to apply a `role`, it seems reasonable to be allowed to override it with an explicit `role`, imo.

**Comment by scottaohara** [comment]
@zcorpan yes i would agree

going with 2, status quo, is not great.  presently browsers are not all in alignment with what's in the HTML spec. Browser-specific (also AT specific) heuristics are a problem, as it means authors don't know what to do unless they know the ins and outs of each browser/AT combo.  that is something i find to be incredibly problematic.

Agreed that the complete removal of heuristics (4) would be problemattic - but also based on recent [testing I've been doing this weekend](https://scottaohara.github.io/tests/html-tables/expose-table.html) (partial - more test cases are being worked on), again browsers / screen readers are not all in alignment here, and that's just beyond frustrating at this point.  The fact that Android Chrome/Firefox with TalkBack are being quite liberal with their exposing of tables in the test cases... well, I'll acknowledge that those were choices that were made.  

Re: (3).  Yes.  I would say the HTML spec has a good outline here. But the noncommittal "probably" statements would be what could be better committed to (in this spec, or HTML, or both. let's not be strangers :) ).
- if role=presentation/none - **layout table**. This is explicit - it's not a "probably".
- if use of caption, th, thead, header scope, border - it's **tabular data**.  In the rare case it's not - eh!  better to expose as a table for the majority of the time rather than being wrong when authors are truly messing up.
- if using the summary attribute - **tabular data** (it's unfortunate this was set to obsolete. it's a useful attribute. this should be re-examined. I question if accessibility concerns were seriously considered with its current status.)
- address how CSS can muck up tables when changing their `display` property value.  

Regarding (1), there's nothing "non-conforming" about using `role=table` on a `table`.  it's a warning because it _should_ be unnecessary when properly using the element and its children.  But it is not an error.  So nothing to address here.

**Comment by dd8** [comment]
I still have a concern on warning about `role=table` for 2 reasons:

1) Heuristics are approximations - so even with a standardised set of heuristics there will still be cases where they produce the wrong result and you need `role=table`. It might be possible to turn the heuristics into a logically certain algorithm that is 100% reliable at detecting layout tables, but given the complexities and history this is going to be difficult.
2) Some organisations have quality standards like "there must be no conformance errors or warnings", which forces devs to remove `role=table` even if that causes harm. 

@scottaohara Might be worth running this by David Sloan - he researched this type of problem in his PhD thesis. I think the example he gave was an author who had removed all heading elements on a page to fix a warning about headings not being nested correctly.
